packages/opentelemetry-instrumentation-fetch/src/fetch.ts

```diff
@@ -378,25 +389,34 @@ export class FetchInstrumentation extends InstrumentationBase<
         ) {
           try {
             endSpanOnError(span, error);
+          } catch(e) {
+            plugin._diag.error('Instrumentation error', e);
           } finally {
             reject(error);
           }
         }
 
         return new Promise((resolve, reject) => {
-          return api.context.with(
-            api.trace.setSpan(api.context.active(), createdSpan),
-            () => {
-              plugin._addHeaders(options, url);
-              plugin._tasksCount++;
-              return original
-                .apply(this, [url, options])
-                .then(
-                  (onSuccess as any).bind(this, createdSpan, resolve),
-                  onError.bind(this, createdSpan, reject)
-                );
-            }
-          );
+          try {
+            return api.context.with(
+              api.trace.setSpan(api.context.active(), createdSpan!),
+              () => {
+                plugin._addHeaders(options, url);
+                plugin._tasksCount++;
+                return original
+                  .apply(this, [url, options])
+                  .then(
+                    (onSuccess as any).bind(this, createdSpan, resolve),
+                    onError.bind(this, createdSpan!, reject)
+                  );
+              }
+            );
+          } catch (e) {
```

Comment:
The only way this can throw is if `original.apply` throws. In your catch block, you are calling it again. This is likely to cause even more problems I think.

Reply:
I wasn't saying the whole catch was bad, just pointing out the control flow error. I don't think this can be merged while the control flow error persists.

Reply:
In theory everything that runs in browser can throw even OTEL API. This try catch here is mostly to capture possible errors thrown in `addHeaders` and whatever new code is added in the future. In XHR it was easier to wrap OTEL API calls so I tried to do it here also which resulted in this ugly construct. 

About the control flow error, it seems all `original.apply` errors are handled in promise and this catch won't fire a second time. I tried to force it with no network, [fetch('http://user:password@example.com')](https://developer.mozilla.org/en-US/docs/Web/API/Request/Request#errors), incorrect options etc. which all throw in promise. I am not sure how to make it throw in a way that this catch captures it.

I can move the try catch to just cover `addHeaders` just to be safe.